### PR TITLE
Apply #pragma pack_matrix at the right token-level granularity

### LINKED ISSUE #1798: DXC does not apply #pragma pack_matrix at the same granularity as FXC
Repro:

```
#pragma pack_matrix(column_major)
typedef
#pragma pack_matrix(row_major)
int2x2
#pragma pack_matrix(column_major)
i22;

void main(out i22 mat : OUT) { mat = i22(11, 12, 21, 22); }
```

DXC output (column major):

```
call void @dx.op.storeOutput.i32(i32 5, i32 0, i32 0, i8 0, i32 11)
call void @dx.op.storeOutput.i32(i32 5, i32 0, i32 0, i8 0, i32 21)
call void @dx.op.storeOutput.i32(i32 5, i32 0, i32 0, i8 0, i32 12)
call void @dx.op.storeOutput.i32(i32 5, i32 0, i32 0, i8 0, i32 22)
```

FXC output (row major):

```
mov o0.xy, l(11,12,0,0)
mov o1.xy, l(21,22,0,0)
```

Even more pedantically, this exhibits the same difference:

```
#pragma pack_matrix(column_major)
typedef matrix<int, 2, 2
#pragma pack_matrix(row_major)
>
#pragma pack_matrix(column_major)
i22;

void main(out i22 mat : OUT) { mat = i22(11, 12, 21, 22); }
```

### LINKED ISSUE #1805: dxc crashes when trying to fill typeloc for matrix orientation attribute
```
#pragma pack_matrix (row_major)
uniform float4x4 gModelViewProjection;
float4 main(float4 a : A) : SV_Position
{
 return  mul(a, gModelViewProjection);
}
```

### PR DESCRIPTION
**Pull request by MrTrillian**
This ensures that #pragma pack_matrix is applied as close to as possible to the same token-level granularity as FXC. The implementation reshuffled the pragma-handling code and in the end is a bit similar to what we had. My initial goal was to ensure that all matrices are marked with a packing orientation annotation, so we can stop relying on `bDefaultRowMajor` everywhere (#1773), but the problem is that diagnostics then start referring to `column_major int2x2` types where there was only `int2x2` in the code. This is still the case once a #pragma has been used though, and is necessary for rewriting while preserving semantics (since the rewriter doesn't rewrite #pragmas). For now I'll stop here and we can reconsider annotating all matrices as part of #1773.

Fixes #1798 
Fixes #1805

### CODE REVIEW

File: tools/clang/lib/Sema/SemaType.cpp [hunk 1]
```diff
@@ -4538,11 +4531,9 @@ static void fillAttributedTypeLoc(AttributedTypeLoc TL,
 
   // HLSL changes begin
   // Don't fill the location info for matrix orientation attributes
-  if (!attrs && !DeclAttrs) {
```
Comment (vcsharma) [inline tools/clang/lib/Sema/SemaType.cpp]: This change should fix #1805.

Comment (MrTrillian) [reply tools/clang/lib/Sema/SemaType.cpp]: Oh nice, that's some real-time bug fixing! Thanks :)

File: tools/clang/test/CodeGenHLSL/quick-test/matrix_orientation_pragma_granularity.hlsl [hunk 2]
```diff
@@ -0,0 +1,24 @@
+// RUN: %dxc /T vs_6_0 /E main > %s | FileCheck %s
```
Comment (vcsharma) [inline tools/clang/test/CodeGenHLSL/quick-test/matrix_orientation_pragma_granularity.hlsl]: Could you add some test scenarios surrounding usage of `/Zpr` and `/Zpc`. AFAIK, these flags should override any type attributes defined files using means such as pragma pack_matrix , etc. ?

File: tools/clang/lib/Sema/SemaType.cpp [hunk 3]
```diff
@@ -4538,11 +4531,9 @@ static void fillAttributedTypeLoc(AttributedTypeLoc TL,
...
-    if (TL.getAttrKind() == AttributedType::attr_hlsl_row_major ||
-        TL.getAttrKind() == AttributedType::attr_hlsl_column_major)
-      return;
-  }
+  if (TL.getAttrKind() == AttributedType::attr_hlsl_row_major ||
```
[the remark this answers is not in the document]
Comment (vcsharma) [reply tools/clang/lib/Sema/SemaType.cpp]: That's right. We should not fill type loc for matrix orientation attributes regardless of `attrs` is null or not.

File: tools/clang/lib/Sema/SemaAttr.cpp [hunk 4]
```diff
@@ -266,13 +266,10 @@ void Sema::ActOnPragmaPack(PragmaPackKind Kind, IdentifierInfo *Name,
 }
 
 void Sema::ActOnPragmaPackMatrix(bool bRowMajor, SourceLocation PragmaLoc) {
-  if (bRowMajor) {
-    PackMatrixRowMajorPragmaOn = true;
-    PackMatrixColMajorPragmaOn = false;
```
[the remark this answers is not in the document]
Comment (MrTrillian) [reply tools/clang/lib/Sema/SemaAttr.cpp]: Nope, hence their removal in Sema.h! :)

File: tools/clang/lib/Sema/SemaHLSL.cpp [hunk 5]
```diff
@@ -11279,56 +11279,6 @@ void Sema::TransferUnusualAttributes(Declarator &D, NamedDecl *NewDecl) {
         D.UnusualAnnotations.size()));
     D.UnusualAnnotations.clear();
   }
-  // pragma pack_matrix.
```
[the remark this answers is not in the document]
Comment (MrTrillian) [reply tools/clang/lib/Sema/SemaHLSL.cpp]: Almost, but you were missing the default annotation for local variables in the parser, so this case was still taking care of that I think.

File: tools/clang/test/CodeGenHLSL/quick-test/matrix_orientation_pragma_granularity.hlsl (hunk 2 above)
Comment (MrTrillian) [reply tools/clang/test/CodeGenHLSL/quick-test/matrix_orientation_pragma_granularity.hlsl]: That's not correct, `/Zpr` and `/Zpc` have exactly the same behavior as having a corresponding `#pragma pack_matrix` right before the start of the file (except as far a rewriting goes). Let me see if I can improve a test to reflect this.

Comment (MrTrillian) [reply tools/clang/test/CodeGenHLSL/quick-test/matrix_orientation_pragma_granularity.hlsl]: Actually if you don't mind, I'll do it as a subsequent checkin. I want to update `matrix_orientation_overrides.hlsl` to have `/Zpr` and `/Zpc` flavors, but that file depends on my other pull request.

Comment (MrTrillian) [reply tools/clang/test/CodeGenHLSL/quick-test/matrix_orientation_pragma_granularity.hlsl]: Here you go: #1807

**Pull request merged**